memory/iter.go

```diff
@@ -7,210 +7,174 @@ import (
 	"github.com/tidwall/btree"
 )
 
-type dsPrefixIter struct {
-	ctx     context.Context
+type iterator struct {
 	db      *Datastore
 	version uint64
-	it      *baseIterator
-	prefix  []byte
-	reverse bool
-	curItem *dsItem
-}
+	it      btree.IterG[dsItem]
 
-type dsRangeIter struct {
-	ctx     context.Context
-	db      *Datastore
-	version uint64
-	it      *baseIterator
-	start   []byte
-	end     []byte
-	reverse bool
-	curItem *dsItem
-}
+	// The key at which this iterator begins.
+	//
+	// This is inclusive only if `mayExactlyMatchStart` is true.
+	start []byte
 
-func newPrefixIter(ctx context.Context, db *Datastore, prefix []byte, reverse bool, version uint64) *dsPrefixIter {
-	pIter := &dsPrefixIter{
-		ctx:     ctx,
-		db:      db,
-		version: version,
-		it:      newBaseIterator(db.values, reverse, prefix, bytesPrefixEnd(prefix)),
-		prefix:  prefix,
-		reverse: reverse,
-	}
-	if reverse {
-		pIter.Seek(bytesPrefixEnd(prefix))
-	} else {
-		pIter.Seek(prefix)
-	}
+	// The key at which this iterator ends, inclusive.
+	end []byte
 
-	// if the first key is an exact match to the prefix, skip next
-	// since prefix is a *strict* subset prefix
-	if pIter.curItem != nil && bytes.Equal(pIter.Key(), prefix) {
-		pIter.Next()
-	}
+	// If this is true, `start` is inclusive, else `start` is exclusive.
+	mayExactlyMatchStart bool
 
-	return pIter
-}
+	// If true, the iterator will iterate in reverse order, from the largest
+	// key to the smallest.
+	reverse bool
 
-func (iter *dsPrefixIter) Domain() (start []byte, end []byte) {
-	return iter.prefix, iter.prefix
+	// hasItem mutates as the iterator progresses, it will be true
+	// if there are remaining items that can iterated through.
+	hasItem bool
 }
 
-func (iter *dsPrefixIter) Valid() bool {
-	return validForPrefix(iter.curItem, iter.prefix)
-}
+func newPrefixIter(db *Datastore, prefix []byte, reverse bool, version uint64) *iterator {
+	start := prefix
+	end := bytesPrefixEnd(prefix)
 
-func validForPrefix(item *dsItem, prefix []byte) bool {
-	if item == nil {
-		return false
+	it := db.values.Iter()
+	var hasItem bool
+	if reverse {
+		hasItem = it.Last()
+	} else {
+		hasItem = it.First()
 	}
 
-	return bytes.HasPrefix(item.key, prefix) && !bytes.Equal(item.key, prefix)
-}
+	iter := &iterator{
+		db:      db,
+		version: version,
+		it:      it,
+		start:   start,
+		end:     end,
+		// A prefix iterator should not return a key exactly matching itself.
```

Comment:
question: What is meant by a `prefix` iterator here?

question: Why should (should be this be a "must"?) it not return a key exactly matching itself.

Reply:
> What is meant by a prefix iterator here?

An iterator that covers a prefix.  I consider this to be covered by the function name, and the only place where it is called.

> must

I like this, thanks - I'll change it to `must`

- [ ] Reword should to must